File: src/515. Find Largest Value in Each Tree Row/515.py
# Definition for a binary tree node.
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def largestValues(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        hashMap = {}

        def dfs(node, depth):
            if not node:
                return
            else:
                if depth not in hashMap:
                    hashMap[depth] = node.val
                else:
                    hashMap[depth] = max(node.val, hashMap[depth])
            dfs(node.left, depth + 1)
            dfs(node.right, depth + 1)
        dfs(root, 1)
        return [hashMap[i] for i in hashMap]

File: src/515. Find Largest Value in Each Tree Row/515.go
func largestValues(root *TreeNode) []int {
    res := make([]int, 0)
    if root == nil {
        return res
    }
    dfsLargestValues(root, &res, 0)
    return res
}


func dfsLargestValues(root *TreeNode, res *[]int, level int) {
    if root == nil {
        return
    }
    if len(*res) == level {
        *res = append(*res, math.MinInt32)
    }
    (*res)[level] = int(math.Max(float64((*res)[level]), float64(root.Val)))
    if root.Left != nil {
        dfsLargestValues(root.Left, res, level+1)
    }
    if root.Right != nil {
        dfsLargestValues(root.Right, res, level+1)
    }
}



func largestValues(root *TreeNode) []int {
	res := make([]int, 0)
	if root == nil {
		return res
	}
	stack := make([]*TreeNode, 0)
	stack = append(stack, root)
	for len(stack) != 0 {
		size := len(stack)
		levelMax := math.MinInt32
		for i := 0; i < size; i++ {
			curNode := stack[0]
			stack = stack[1:]
			levelMax = int(math.Max(float64(curNode.Val), float64(levelMax)))
			if curNode.Left != nil {
				stack = append(stack, curNode.Left)
			}
			if curNode.Right != nil {
				stack = append(stack, curNode.Right)
			}
		}
		res = append(res, levelMax)
	}
	return res
}